crates/autopilot/src/run_loop.rs

```diff
@@ -560,45 +557,45 @@ impl RunLoop {
                     );
                     None
                 }
+            });
+
+        // Winners are selected one by one, starting from the best solution,
+        // until `max_winners_per_auction` are selected. The solution is a winner
+        // if it swaps tokens that are not yet swapped by any other already
+        // selected winner.
+        let mut already_swapped_tokens = HashSet::new();
+        let mut winners = 0;
+        let solutions = solutions
+            .map(|participant| {
+                let swapped_tokens = participant
+                    .solution
+                    .orders()
+                    .iter()
+                    .flat_map(|(_, order)| vec![order.sell.token, order.buy.token])
+                    .collect::<HashSet<_>>();
+
+                let is_winner = swapped_tokens.is_disjoint(&already_swapped_tokens)
```

Comment:
One more safety precaution we might have to take here is to make sure no 2 orders trade WETH or ETH as they are generally considered to be equivalent but not in this check.

Reply:
This is actually a very nice suggestion for improvement. 

Added a separate task to not lose time on it right now since I'd like this PR merged asap.
https://github.com/cowprotocol/services/issues/3043https://github.com/cowprotocol/services/issues/3043